### PR TITLE
Kombu consumer error handling

### PR DESCRIPTION
**Pull request by a-belevich**
Closes # 

## 💸 TL;DR
Increase a choice of error handling strategies in the KombuConsumer. In addition to writing their own error handling function, or relying on standard behavior that requeues message on error, it adds a 3rd option: republish the message. I.e., acknowledge the current message and publish its copy back into the queue, with incremented retries counter (similar to what Celery is doing).

## 📜 Details
[Design Doc](https://docs.google.com/document/d/1FP7pdZRBh4ULpCdM0t8TJwDe3K6oA6R2rM56vA66Kf0/edit?usp=sharing)

## 🧪 Testing Steps / Validation
The republish mode with retries limit was tested on Snoodev. Unit tests were added..

## ✅ Checks

- [ ] CI tests (if present) are passing
- [ ] Adheres to code style for repo
- [ ] Contributor License Agreement (CLA) completed if not a Reddit employee

### CODE REVIEW

File: baseplate/frameworks/queue_consumer/kombu.py
Comment (nsheaps) [inline baseplate/frameworks/queue_consumer/kombu.py:56]: are there associated spec changes for these new metrics?

Comment (a-belevich) [reply baseplate/frameworks/queue_consumer/kombu.py:56]: Sorry, I don't understand; which spec are we talking about? Any reference or example to look at, please?

Comment (a-belevich) [reply baseplate/frameworks/queue_consumer/kombu.py:56]: Ohh,  you probably meant tests? Sure, will do; I only managed to run it in snoodev today, the tests are next.

Comment (a-belevich) [reply baseplate/frameworks/queue_consumer/kombu.py:56]: Added unit tests.

**Review by nsheaps — CHANGES_REQUESTED** [review]
I appreciate the effort you've put into this and thank you for providing the design doc. However, this is a change to the metrics emitted from baseplate, which has to be documented and agreed up on first in [github enterprise]/reddit/baseplate.spec, and your design doc doesn't even list the baseplate team as a stakeholder nor does it have signoff. The spec change and signoff are important, since changes like these need docs updates, and code updates in other baseplate implementations for consistency.

Please start with a baseplate.spec change and attach your design doc there (you can leave this open in the mean time). Once that's buttoned up including approvals we can continue with this PR.

Additionally, please don't remove reviewers just because one person started to review, we tag two people for a reason.

**Comment by a-belevich** [comment]
Thank you, will do.

**Review by KTAtkinson — CHANGES_REQUESTED** [review]
+1 to Nate's comment. please remember that proposals for changes to Baseplate should always be reviewed by the Baseplate owners.

**Comment by nsheaps** [comment]
I'm approving the run of CI/CD just so you can get some notion of pass/fail, but the PR will remain in this state till the spec change can be approved. Thanks for being flexible with that! Happy to discuss this further in the #baseplate-spec channel

**Comment by a-belevich** [comment]
https://github.snooguts.net/reddit/baseplate.spec/pull/70 is the spec PR; it's approved by I have no permissions to merge it.

File: baseplate/frameworks/queue_consumer/kombu.py
Comment (KTAtkinson) [inline baseplate/frameworks/queue_consumer/kombu.py:264]: This applies to all messages not just AMQP ones. Is this intentional?

File: baseplate/frameworks/queue_consumer/kombu.py
Comment (KTAtkinson) [inline baseplate/frameworks/queue_consumer/kombu.py:298]: This function only gets called if the error is not associated with the message handle function. Is this intended?

File: baseplate/frameworks/queue_consumer/kombu.py
Comment (KTAtkinson) [inline baseplate/frameworks/queue_consumer/kombu.py:301]: What sort of error will be considered fatal?

File: baseplate/frameworks/queue_consumer/kombu.py
Comment (KTAtkinson) [inline baseplate/frameworks/queue_consumer/kombu.py:201]: Having the queue always re-enqueue failed messages opens the system up to infinite retries. Rather than retrying failed messages infinitely we usually suggest using dead-letter queues. I would consider infinite retries to be actively harmful to the system.

File: baseplate/frameworks/queue_consumer/kombu.py
Comment (a-belevich) [reply baseplate/frameworks/queue_consumer/kombu.py:264]: Yes, it's intentional. Literally 12 lines below:  
```
AMQP_ACTIVE_MESSAGES.labels(
                **prometheus_labels._asdict()
            ).track_inprogress()
```  
Or, a bit more below:
```
            AMQP_PROCESSED_TOTAL.labels(
                **prometheus_labels._asdict(), amqp_success=prometheus_success
            ).inc()
```
I.e. as far as I can see this change extends an existing functionality, and not creating new conventions.

File: baseplate/frameworks/queue_consumer/kombu.py
Comment (a-belevich) [reply baseplate/frameworks/queue_consumer/kombu.py:201]: It's a pre-existing behavior which makes my changes backward compatible; in current version it looks like
```
            if self.error_handler_fn:
                self.error_handler_fn(context, message_body, message, exc)
            else:
                message.requeue()
```

File: baseplate/frameworks/queue_consumer/kombu.py
Comment (a-belevich) [reply baseplate/frameworks/queue_consumer/kombu.py:298]: Yes, it's intended, and it's a pre-existing behavior:
```
            if self.error_handler_fn:
                self.error_handler_fn(context, message_body, message, exc)
            else:
                message.requeue()
```
I.e. behavior, as described in the design doc, is extended. Before this change users had 2 options: to write their custom error handler, or to rely on default strategy of infinite retries. Now they have 3 options: custom error handler, configurable republishing mechanism (that can support retries limiting) or the old default.

File: baseplate/frameworks/queue_consumer/kombu.py
Comment (a-belevich) [reply baseplate/frameworks/queue_consumer/kombu.py:301]: It's a pre-existing behavior; I have no idea which error any user of this framework can consider fatal.

File: baseplate/frameworks/queue_consumer/kombu.py
Comment (KTAtkinson) [reply baseplate/frameworks/queue_consumer/kombu.py:201]: Thanks for pointing that out! That's unfortunate, but you are right that we need to be backwards compatible.

**Review by nsheaps — APPROVED** [review]

**Review by KTAtkinson — APPROVED** [review]

**Pull request merged**
